File: Source/Client-HGE/DxComponent/DxImageProgress.pas
unit DxImageProgress;

interface

uses
  Windows,
  Types,
  Classes,
  Controls,
  SysUtils,
  Graphics,
  HGE,
  HGECanvas,
  DxComponents,
  DxControls,
  GameImages,
  Math,
  HGEFontEx;

type
  TDxImageProgress = class;

  TProgressSetting = class(TPersistent)
  private
    FOwner:TDxImageProgress;

    FFont:TDxFont;

    FOnChange:TNotifyEvent;
    FOnGetImage:TOnGetImage;

    FImageType:TImageType; // ??WIL??
    FImage:TGameImages;

    FImageBG:Integer;
    FImageProgress:Integer;
    FImageProgressX:Integer;
    FImageProgressY:Integer;

    FValueType:TProgressValueType;
    FValueSplite:string;

    FValueAlignment:TAlignment;
    FValuePrefix:string;
    FValueSuffix:string;

    FMax:LongWord;
    FMin:LongWord;
    FValue:LongWord;

    procedure SetImageType(Value:TImageType);
    procedure SetOnGetImage(Value:TOnGetImage);
    procedure SetImageBG(const Value:Integer);
    procedure SetImageProgress(const Value:Integer);

    procedure SetFont(Value:TDxFont);
    procedure FontChange(Sender:TObject); stdcall;
  protected
    procedure Changed; // dynamic;
  public
    constructor Create(AOwner:TDxImageProgress);
    destructor Destroy; override;
    procedure Assign(Source:TPersistent); override;

    property Image:TGameImages read FImage;
    property OnChange:TNotifyEvent read FOnChange write FOnChange;
    property OnGetImage:TOnGetImage read FOnGetImage write SetOnGetImage;
  published
    property Font:TDxFont read FFont write SetFont;

    property ImageType:TImageType read FImageType write SetImageType;
    property ImageBG:Integer read FImageBG write SetImageBG;
    property ImageProgress:Integer read FImageProgress write SetImageProgress;

    property ImageProgressX:Integer read FImageProgressX write FImageProgressX;
    property ImageProgressY:Integer read FImageProgressY write FImageProgressY;

    property ValueType:TProgressValueType read FValueType write FValueType;
    property ValueSplite:string read FValueSplite write FValueSplite;
    property ValueAlignment:TAlignment read FValueAlignment write FValueAlignment;
    property ValuePrefix:string read FValuePrefix write FValuePrefix;
    property ValueSuffix:string read FValueSuffix write FValueSuffix;

    property Max:LongWord read FMax write FMax;
    property Min:LongWord read FMin write FMin;
    property Value:LongWord read FValue write FValue;
  end;

  TDxImageProgress = class(TDxControl)
  private
    FProgressSetting:TProgressSetting;

    procedure RecallAutoSize;
  protected
    procedure DoOnInRealArea(X, Y:Integer; var IsRealArea:Boolean); override;
    procedure SetOnGetImage(Value:TOnGetImage); override;
  public
    constructor Create(AOwner:{$IF  CLIENTEXE = 1}TDxControl{$ELSE}TComponent{$IFEND}); override;
    destructor Destroy; override;
    procedure Assign(Source:TDxControl); override;
    procedure Paint; override;
  published
    property ProgressSetting:TProgressSetting read FProgressSetting write FProgressSetting;
    property AutoSize;
  end;

implementation

constructor TProgressSetting.Create(AOwner:TDxImageProgress);
begin
  inherited Create;
  FOwner := AOwner;

  FFont := TDxFont.Create;
  FFont.Bold := True;
  FFont.Color := clWhite;
  FFont.OnChange := FontChange;

  FImageType := Prguse_wil;

  FOwner := AOwner;
  FImageBG := -1;
  FImageProgress := -1;

  FImageProgressX := 0;
  FImageProgressY := 0;

  FValueType := vtValue;
  FValueSplite := '-';
  FValueAlignment := taCenter;
  FValuePrefix := '';
  FValueSuffix := '';

  FMax := 100;
  FMin := 0;
  FValue := 50;

  FImage := nil;
  FOnChange := nil;
  FOnGetImage := nil;
end;

destructor TProgressSetting.Destroy;
begin
  FFont.Free;
  inherited;
end;

procedure TProgressSetting.SetImageType(Value:TImageType);
begin
  if FImageType <> Value then begin
    FImageType := Value;
    if Assigned(FOnGetImage) then
      FOnGetImage(Self, FImageType, TObject(FImage));
    Changed;
  end;
end;

procedure TProgressSetting.SetOnGetImage(Value:TOnGetImage);
begin
  FOnGetImage := Value;
  if Assigned(FOnGetImage) then
    FOnGetImage(Self, FImageType, TObject(FImage));
  Changed;
end;

procedure TProgressSetting.SetFont(Value:TDxFont);
begin
  FFont.Assign(Value);
end;

procedure TProgressSetting.FontChange(Sender:TObject); stdcall;
begin
  if Assigned(FOnChange) then
    FOnChange(Self);
end;

procedure TProgressSetting.SetImageBG(const Value:Integer);
begin
  if FImageBG <> Value then begin
    FImageBG := Value;
    Changed;
  end;
end;

procedure TProgressSetting.SetImageProgress(const Value:Integer);
begin
  if FImageProgress <> Value then begin
    FImageProgress := Value;
    Changed;
  end;
end;

procedure TProgressSetting.Changed;
begin
  FOwner.RecallAutoSize;

  if Assigned(FOnChange) then
    FOnChange(Self);
end;

procedure TProgressSetting.Assign(Source:TPersistent);
begin
  if Source is TProgressSetting then begin
    OnGetImage := TProgressSetting(Source).OnGetImage;
    FImageType := TProgressSetting(Source).FImageType;

    FFont.Assign(TProgressSetting(Source).FFont);

    FImageBG := TProgressSetting(Source).FImageBG;
    FImageProgress := TProgressSetting(Source).FImageProgress;

    FValueType := TProgressSetting(Source).FValueType;
    FValueSplite := TProgressSetting(Source).FValueSplite;
    FValueAlignment := TProgressSetting(Source).FValueAlignment;
    FValuePrefix := TProgressSetting(Source).FValuePrefix;
    FValueSuffix := TProgressSetting(Source).FValueSuffix;

    FMax := TProgressSetting(Source).FMax;
    FMin := TProgressSetting(Source).FMin;
    FValue := TProgressSetting(Source).FValue;

    Changed;
  end;
end;

{------------------------------------------------------------------------------}

constructor TDxImageProgress.Create(AOwner:{$IF  CLIENTEXE = 1}TDxControl{$ELSE}TComponent{$IFEND});
begin
  inherited Create(AOwner);

  FProgressSetting := TProgressSetting.Create(Self);
  FProgressSetting.OnGetImage := OnGetImage;

  Width := 90;
  Height := 20;
end;

destructor TDxImageProgress.Destroy;
begin
  FProgressSetting.Free;
  inherited Destroy;
end;

procedure TDxImageProgress.Paint;
var
  I, nWidth, nHeight:Integer;
  D:TTexture;
  vtRect, vbRect, SrcRect:TRect;
  S:string;
  HGEFont:THGEFont;
  TextImages:TImageInfos;
begin
  vbRect := VisibleRect;
  if (vbRect.Bottom <= vbRect.Top) or (vbRect.Right <= vbRect.Left) then Exit;
  vtRect := VirtualRect;

  if FProgressSetting.Image <> nil then begin
    if (FProgressSetting.FImageBG >= 0) then begin
      D := FProgressSetting.Image.Images[FProgressSetting.FImageBG];

      if D <> nil then begin
        SrcRect := Rect(0, 0, Min(D.Width, Width), Min(D.Height, Height));
        GameCanvas.Draw(vtRect.Left, vtRect.Top, SrcRect, D);
      end;
    end;

    if (FProgressSetting.FImageProgress >= 0) and
      (FProgressSetting.FMax > FProgressSetting.FMin) and
      (FProgressSetting.FValue >= FProgressSetting.Min) then begin
      D := FProgressSetting.Image.Images[FProgressSetting.FImageProgress];

      if D <> nil then begin
        SrcRect := Rect(0, 0, D.Width, D.Height);
        SrcRect.Right := SrcRect.Left + Round((SrcRect.Right - SrcRect.Left) / (FProgressSetting.FMax - FProgressSetting.FMin) * Min(FProgressSetting.FValue, FProgressSetting.FMax));
        GameCanvas.Draw(vtRect.Left + FProgressSetting.FImageProgressX, vtRect.Top + FProgressSetting.FImageProgressY, SrcRect, D);
      end;
    end;

    if (FProgressSetting.FImageBG >= 0) and
      (FProgressSetting.FMax > FProgressSetting.FMin) and
      (FProgressSetting.FValue >= FProgressSetting.Min) then begin
      if FProgressSetting.FValueType = vtValueAndMax then
        S := FProgressSetting.ValuePrefix + IntToStr(FProgressSetting.FValue) + FProgressSetting.FValueSplite + IntToStr(FProgressSetting.FMax) + FProgressSetting.FValueSuffix
      else if FProgressSetting.FValueType = vtValue then
        S := FProgressSetting.ValuePrefix + IntToStr(FProgressSetting.FValue) + FProgressSetting.FValueSuffix
      else if FProgressSetting.FValueType = vtPercentage then
        S := FProgressSetting.ValuePrefix + Format('%d%%', [Round(FProgressSetting.FValue / (FProgressSetting.FMax - FProgressSetting.FMin) * 100)]) + FProgressSetting.FValueSuffix
      else
        S := '';

      if Length(S) > 0 then begin
        HGEFont := TextureFonts.FindFont(FProgressSetting.FFont.Name, FProgressSetting.FFont.Size, FProgressSetting.FFont.Style);

        if HGEFont <> nil then begin
          TextImages := HGEFont.GetImageInfos(S);

          nWidth := 0;
          nHeight := 0;
          for I := 0 to Length(TextImages) - 1 do begin
            if nWidth < TextImages[I].Width then
              nWidth := TextImages[I].Width;

            if TextImages[I].Height <= 0 then
              Inc(nHeight, HGEFont.TextHeight('0'))
            else
              Inc(nHeight, TextImages[I].Height);
          end;

          if FProgressSetting.Font.Bold then begin
            Inc(nWidth, 2);
            Inc(nHeight, 2);
          end;

          DrawCaption(HGEFont, FProgressSetting.Font, FProgressSetting.FValueAlignment, TextImages, Bounds(vtRect.Left, vtRect.Top, nWidth, nHeight), vbRect, vtRect, 0, 0);
        end;
      end;
    end;
  end;

  if Assigned(OnStartSubPaint) then
    OnStartSubPaint(Self);

  for I := ControlCount - 1 downto 0 do begin
    if not Visible then break;
    if (Control[I].Visible) then
      Control[I].Paint;
  end;

  if Visible and Assigned(OnStopPaint) then
    OnStopPaint(Self);

  if Designing then begin
    FrameRect(vtRect, vtRect, vbRect, BorderColor.Up.Color);
  end;
end;

procedure TDxImageProgress.SetOnGetImage(Value:TOnGetImage);
begin
  inherited;
  FProgressSetting.OnGetImage := Value;
end;

procedure TDxImageProgress.DoOnInRealArea(X, Y:Integer; var IsRealArea:Boolean);
begin
  inherited;
end;

procedure TDxImageProgress.RecallAutoSize;
var
  D:TTexture;
begin
  if not AutoSize then Exit;

  if FProgressSetting.Image <> nil then begin
    if (FProgressSetting.FImageBG >= 0) then begin
      D := FProgressSetting.Image.Images[FProgressSetting.FImageBG];

      if (D <> nil) and (D.Width * D.Height > 4) then begin
        Width := D.Width;
        Height := D.Height;
      end;
    end;
  end;
end;

procedure TDxImageProgress.Assign(Source:TDxControl);
begin
  if Source is TDxImageProgress then begin
    Left := Source.Left;
    Top := Source.Top;
    Width := Source.Width;
    Height := Source.Height;
    Enabled := Source.Enabled;
    Visible := Source.Visible;

    Transparent := Source.Transparent; // ????
    EnableFocus := Source.EnableFocus; // ????????
    Floating := Source.Floating; // ??????
    OwnerMove := Source.OwnerMove;

    MouseEvents := Source.MouseEvents;

    //Designing := Source.Designing;                                                                    // ??????
    ReferenceX := Source.ReferenceX;
    AdjustYByHeight := Source.AdjustYByHeight;
    TopAlignment := Source.TopAlignment;

    Center := Source.Center;
    Align := Source.Align;
    AutoSize := Source.AutoSize;
    BackgroundColor := Source.BackgroundColor;

    ImageIndex.Assign(Source.ImageIndex);
    BorderColor.Assign(Source.BorderColor);

    FProgressSetting.Assign(TDxImageProgress(Source).FProgressSetting);
  end;
end;

end.

File: GXX.CSharp/src/GXX.Client/DxComponent/DxImageProgress.cs
using System;
using System.Collections.Generic;

namespace GXX.Client.DxComponent;

/// <summary>
/// DxImageProgress.pas 1:1 逐字移植（1-402）。
///
/// 覆盖两类内容：
///   1. TProgressSetting（23-87 / 109-225）—— 设置对象 + Changed 通知链
///   2. TDxImageProgress（89-105 / 229-400）—— 控件：构造默认值、进度几何、值文本、RecallAutoSize
///
/// 保真要点（原文易错处，逐条固化）：
///   * 构造默认值（109-141）：Font.Bold=True、Font.Color=clWhite、ImageType=Prguse_wil、
///     ImageBG=-1、ImageProgress=-1、ImageProgressX/Y=0、ValueType=vtValue、ValueSplite='-'
///     （原文拼写，保留）、ValueAlignment=taCenter、ValuePrefix/Suffix=''、
///     Max=100、Min=0、**Value=50**（不是 0）。
///   * SetImageType / SetOnGetImage 都会在**赋值之后**立刻回调一次 OnGetImage（149-165）。
///   * Changed（194-200）：**先 FOwner.RecallAutoSize，再 OnChange**（顺序不可换）。
///   * Assign（202-225）：走 `OnGetImage := ...`（属性写 → 会回调）而 ImageType 是**直写字段**
///     （不触发 OnGetImage）—— 这种不对称是原文如此。
///   * SetFont（167-170）用 FFont.**Assign**(Value)（拷贝而非替换引用；FFont 永不为 nil）。
///   * 控件构造（229-238）：Width=90、Height=20，且把 OnGetImage 接到 FProgressSetting 上。
///   * Paint（246-337）三块绘制的**四条守卫各不相同**（见 PaintTo 逐条注释）。
///   * RecallAutoSize（350-366）：AutoSize 为假直接 Exit；**要求 D.Width * D.Height &gt; 4**
///     （严格大于 4，故 1×1/2×2/1×4 都不触发）。
///
/// 接缝：GameCanvas.Draw / TextureFonts.FindFont / HGEFont（HGECanvas/HGEFontEx 未移植）
///       → IDxSurfacePainter / TDxFontEnv / IDxImageLibrary。
/// </summary>
public sealed class TProgressSetting
{
    /// <summary>DxImageProgress.pas 25 FOwner:TDxImageProgress。</summary>
    public TDxImageProgress Owner { get; internal set; }

    private readonly TDxFont _font = new();
    private TImageType _imageType = TImageType.Prguse_wil;
    private int _imageBG = -1;
    private int _imageProgress = -1;
    private int _imageProgressX;
    private int _imageProgressY;
    private TProgressValueType _valueType = TProgressValueType.vtValue;
    private string _valueSplite = "-";
    private TDxAlignment _valueAlignment = TDxAlignment.taCenter;
    private string _valuePrefix = "";
    private string _valueSuffix = "";
    private uint _max = 100;
    private uint _min;
    private uint _value = 50;

    /// <summary>DxImageProgress.pas 29 FOnChange（TNotifyEvent）。</summary>
    public Action<TProgressSetting> OnChange;

    /// <summary>DxImageProgress.pas 30 FOnGetImage（TOnGetImage）。</summary>
    public Action<TProgressSetting, TImageType, IDxImageLibrary> OnGetImage;

    /// <summary>DxImageProgress.pas 109-141 构造函数 1:1。</summary>
    public TProgressSetting(TDxImageProgress owner)
    {
        Owner = owner;

        _font.Bold = true;
        _font.Color = TDxColor.clWhite;
        _font.OnChange = FontChange;

        _imageType = TImageType.Prguse_wil;

        Owner = owner;               // 原文 121：FOwner := AOwner（重复赋值，照抄）
        _imageBG = -1;
        _imageProgress = -1;

        _imageProgressX = 0;
        _imageProgressY = 0;

        _valueType = TProgressValueType.vtValue;
        _valueSplite = "-";
        _valueAlignment = TDxAlignment.taCenter;
        _valuePrefix = "";
        _valueSuffix = "";

        _max = 100;
        _min = 0;
        _value = 50;

        Image = null;
        OnChange = null;
        OnGetImage = null;
    }

    /// <summary>DxImageProgress.pas 65 Image:TGameImages（只读属性）。</summary>
    public IDxImageLibrary Image { get; private set; }

    /// <summary>接缝 SetImage(Value)：原文由 OnGetImage 回调提供。</summary>
    public void SetImage(IDxImageLibrary value) => Image = value;

    /// <summary>DxImageProgress.pas 69 Font:TDxFont（read FFont write SetFont）。</summary>
    public TDxFont Font
    {
        get => _font;
        set => SetFont(value);
    }

    /// <summary>DxImageProgress.pas 71 / 149-157 ImageType。</summary>
    public TImageType ImageType
    {
        get => _imageType;
        set => SetImageType(value);
    }

    /// <summary>DxImageProgress.pas 72 / 178-184 ImageBG（默认 -1）。</summary>
    public int ImageBG
    {
        get => _imageBG;
        set { if (_imageBG != value) { _imageBG = value; Changed(); } }
    }

    /// <summary>DxImageProgress.pas 73 / 186-192 ImageProgress（默认 -1）。</summary>
    public int ImageProgress
    {
        get => _imageProgress;
        set { if (_imageProgress != value) { _imageProgress = value; Changed(); } }
    }

    /// <summary>DxImageProgress.pas 75 ImageProgressX（published 直写字段，无 setter）。</summary>
    public int ImageProgressX { get => _imageProgressX; set => _imageProgressX = value; }

    /// <summary>DxImageProgress.pas 76 ImageProgressY（published 直写字段）。</summary>
    public int ImageProgressY { get => _imageProgressY; set => _imageProgressY = value; }

    /// <summary>DxImageProgress.pas 78 ValueType（published 直写字段，默认 vtValue）。</summary>
    public TProgressValueType ValueType { get => _valueType; set => _valueType = value; }

    /// <summary>DxImageProgress.pas 79 ValueSplite（原文拼写；默认 '-'）。</summary>
    public string ValueSplite { get => _valueSplite; set => _valueSplite = value; }

    /// <summary>DxImageProgress.pas 80 ValueAlignment（默认 taCenter）。</summary>
    public TDxAlignment ValueAlignment { get => _valueAlignment; set => _valueAlignment = value; }

    /// <summary>DxImageProgress.pas 81 ValuePrefix（默认 ''）。</summary>
    public string ValuePrefix { get => _valuePrefix; set => _valuePrefix = value; }

    /// <summary>DxImageProgress.pas 82 ValueSuffix（默认 ''）。</summary>
    public string ValueSuffix { get => _valueSuffix; set => _valueSuffix = value; }

    /// <summary>DxImageProgress.pas 84 Max:LongWord（默认 100；published 直写字段）。</summary>
    public uint Max { get => _max; set => _max = value; }

    /// <summary>DxImageProgress.pas 85 Min:LongWord（默认 0；published 直写字段）。</summary>
    public uint Min { get => _min; set => _min = value; }

    /// <summary>DxImageProgress.pas 86 Value:LongWord（默认 50；published 直写字段）。</summary>
    public uint Value { get => _value; set => _value = value; }

    /// <summary>DxImageProgress.pas 149-157 SetImageType 1:1。</summary>
    public void SetImageType(TImageType value)
    {
        if (_imageType != value)
        {
            _imageType = value;
            if (OnGetImage != null)
                OnGetImage(this, _imageType, Image);
            Changed();
        }
    }

    /// <summary>DxImageProgress.pas 159-165 SetOnGetImage 1:1（赋值后**无条件**回调一次 + Changed）。</summary>
    public void SetOnGetImage(Action<TProgressSetting, TImageType, IDxImageLibrary> value)
    {
        OnGetImage = value;
        if (OnGetImage != null)
            OnGetImage(this, _imageType, Image);
        Changed();
    }

    /// <summary>DxImageProgress.pas 167-170 SetFont：FFont.Assign(Value)（拷贝）。</summary>
    public void SetFont(TDxFont value)
    {
        _font.Assign(value);
    }

    /// <summary>DxImageProgress.pas 172-176 FontChange(stdcall)：只转发 FOnChange。</summary>
    public void FontChange(TDxFont sender)
    {
        OnChange?.Invoke(this);
    }

    /// <summary>DxImageProgress.pas 194-200 Changed 1:1：**先** Owner.RecallAutoSize，**再** OnChange。</summary>
    public void Changed()
    {
        Owner?.RecallAutoSize();

        OnChange?.Invoke(this);
    }

    /// <summary>DxImageProgress.pas 202-225 Assign 1:1。</summary>
    public void Assign(TProgressSetting source)
    {
        if (source is TProgressSetting src)
        {
            // 原文 205：OnGetImage := TProgressSetting(Source).OnGetImage;
            //          （属性写 → 立即回调一次 → 但此时 _imageType 还是旧值）
            SetOnGetImage(src.OnGetImage);
            _imageType = src._imageType;         // 原文 206：FImageType 直写（不回调查询）

            _font.Assign(src._font);             // 原文 208：FFont.Assign

            _imageBG = src._imageBG;
            _imageProgress = src._imageProgress;

            _valueType = src._valueType;
            _valueSplite = src._valueSplite;
            _valueAlignment = src._valueAlignment;
            _valuePrefix = src._valuePrefix;
            _valueSuffix = src._valueSuffix;

            _max = src._max;
            _min = src._min;
            _value = src._value;

            Changed();
        }
    }
}

/// <summary>DxImageProgress.pas 89-105 / 229-400 TDxImageProgress 1:1。</summary>
public sealed class TDxImageProgress : TDxControl
{
    /// <summary>DxImageProgress.pas 91 FProgressSetting。</summary>
    public TProgressSetting ProgressSetting { get; }

    /// <summary>接缝：TGameImages（WIL 图库）—— 原文通过 OnGetImage 注入。</summary>
    public IDxImageLibrary ImageLibrary { get; set; }

    // DFM: TDxImageProgress 无同名 .dfm（DxImageProgress.dfm 在 Source/Client-HGE 全树不存在），
    // 布局属性的设计期取值落在各 .GUI 文件的 TguiProgress 记录里（见报告第 4 项）。

    public TDxImageProgress()
    {
        // DxImageProgress.pas 231-237 逐项
        ProgressSetting = new TProgressSetting(this);
        // 原文 234：FProgressSetting.OnGetImage := OnGetImage;（FProgressSetting 的 TOnGetImage 属性
        // 由控件的 OnGetImage 属性提供；因为 FProgressSetting 刚构造、FOnGetImage 仍为 nil，
        // 这次赋值**不会**触发回调）。此处等价：设置项先负责把图库写回，再转发控件级通知。
        ProgressSetting.OnGetImage = ProgressSettingOnGetImage;

        Width = 90;
        Height = 20;
    }

    /// <summary>ProgressSetting 查询图库时的落点：先写回 ImageLibrary，再转发控件级 OnGetImage。</summary>
    private void ProgressSettingOnGetImage(TProgressSetting sender, TImageType imageType, IDxImageLibrary image)
    {
        sender.SetImage(ImageLibrary);
        ControlOnGetImage?.Invoke(ImageIndex, imageType);
    }

    /// <summary>控件级 OnGetImage 通知的暂存（原文由 TDxControl.OnGetImage 属性持有）。</summary>
    private Action<TDxImageIndex, TImageType> ControlOnGetImage;

    /// <summary>DxImageProgress.pas 339-343 SetOnGetImage：先 inherited 再转发给 ProgressSetting。</summary>
    public void SetOnGetImage(Action<TDxImageIndex, TImageType> value)
    {
        base.SetOnGetImage(value);
        ControlOnGetImage = value;
        ProgressSetting.OnGetImage = ProgressSettingOnGetImage;
    }

    /// <summary>
    /// DxImageProgress.pas 246-337 Paint 的绘制判定 1:1（不落像素，只产出绘制请求）。
    /// 四条守卫彼此独立：
    ///   背景块    ：Image &lt;&gt; nil 且 ImageBG &gt;= 0 且取到图 → Draw(vt.Left, vt.Top, SrcRect, D)
    ///              其中 SrcRect = Rect(0, 0, Min(D.Width, Width), Min(D.Height, Height))
    ///   进度块    ：Image &lt;&gt; nil 且 ImageProgress &gt;= 0 且 Max &gt; Min 且 Value &gt;= Min 且取到图
    ///              SrcRect = Rect(0, 0, D.Width, D.Height) 后 Right := Left +
    ///              Round((Right-Left) / (Max-Min) * Min(Value, Max))
    ///   值文本块  ：**ImageBG &gt;= 0**（不是 ImageProgress！）且 Max &gt; Min 且 Value &gt;= Min
    ///   Designing ：最后 FrameRect(vt, vt, vb, BorderColor.Up.Color)
    /// </summary>
    public TDxImageProgressPaintResult PaintTo(IDxSurfacePainter painter)
    {
        var result = new TDxImageProgressPaintResult();

        var vbRect = VisibleRect;
        if (vbRect.Bottom <= vbRect.Top || vbRect.Right <= vbRect.Left)
        {
            // 原文如此（256）：退化时**连 Designing 边框都不画**，直接 Exit。
            return result;
        }
        var vtRect = VirtualRect;

        Painter = painter ?? Painter;

        if (ProgressSetting.Image != null)
        {
            if (ProgressSetting.ImageBG >= 0)
            {
                var d = ProgressSetting.Image.GetImage(ProgressSetting.ImageBG);
                if (d != null)
                {
                    var srcRect = TDxRect.Rect(0, 0, Math.Min(d.Width, Width), Math.Min(d.Height, Height));
                    Painter.Draw(vtRect.Left, vtRect.Top, srcRect, d);
                    result.BackgroundSrcRect = srcRect;
                    result.DrewBackground = true;
                }
            }

            if (ProgressSetting.ImageProgress >= 0
                && ProgressSetting.Max > ProgressSetting.Min
                && ProgressSetting.Value >= ProgressSetting.Min)
            {
                var d = ProgressSetting.Image.GetImage(ProgressSetting.ImageProgress);
                if (d != null)
                {
                    var srcRect = ComputeProgressSrcRect(d.Width, d.Height, ProgressSetting.Max,
                        ProgressSetting.Min, ProgressSetting.Value);
                    Painter.Draw(vtRect.Left + ProgressSetting.ImageProgressX,
                        vtRect.Top + ProgressSetting.ImageProgressY, srcRect, d);
                    result.ProgressSrcRect = srcRect;
                    result.DrewProgress = true;
                }
            }

            if (ProgressSetting.ImageBG >= 0
                && ProgressSetting.Max > ProgressSetting.Min
                && ProgressSetting.Value >= ProgressSetting.Min)
            {
                string s = ComputeValueText(ProgressSetting.ValueType, ProgressSetting.ValuePrefix,
                    ProgressSetting.ValueSuffix, ProgressSetting.ValueSplite, ProgressSetting.Max,
                    ProgressSetting.Min, ProgressSetting.Value);
                result.ValueText = s;

                if (s.Length > 0)
                {
                    var hgeFont = FindFont(ProgressSetting.Font);
                    if (hgeFont != null)
                    {
                        var textImages = GetImageInfos(hgeFont, s);

                        int nWidth = 0;
                        int nHeight = 0;
                        for (int i = 0; i <= textImages.Count - 1; i++)
                        {
                            if (nWidth < textImages[i].Width)
                                nWidth = textImages[i].Width;

                            if (textImages[i].Height <= 0)
                                nHeight += TextHeight(hgeFont, "0");
                            else
                                nHeight += textImages[i].Height;
                        }

                        if (ProgressSetting.Font.Bold)
                        {
                            nWidth += 2;
                            nHeight += 2;
                        }

                        DrawCaption(hgeFont, ProgressSetting.Font, ProgressSetting.ValueAlignment, textImages,
                            TDxRect.Bounds(vtRect.Left, vtRect.Top, nWidth, nHeight), vbRect, vtRect, 0, 0);
                        result.DrewValueText = true;
                    }
                }
            }
        }

        if (Designing)
        {
            Painter.FrameRect(vtRect, vtRect, vbRect, BorderColor.Up.Color);
            result.DrewDesignFrame = true;
        }

        return result;
    }

    /// <summary>
    /// DxImageProgress.pas 275-277 的进度源矩形 1:1：
    /// SrcRect := Rect(0, 0, D.Width, D.Height);
    /// SrcRect.Right := SrcRect.Left + Round((SrcRect.Right - SrcRect.Left)
    ///                 / (Max - Min) * Min(Value, Max));
    /// 注意除法是**浮点**（/ 而非 div），Round 为银行家舍入，Min 是 Delphi Math.Min(LongWord)。
    /// </summary>
    public static TDxRect ComputeProgressSrcRect(int dWidth, int dHeight, uint max, uint min, uint value)
    {
        var srcRect = TDxRect.Rect(0, 0, dWidth, dHeight);
        srcRect.Right = srcRect.Left + TDxCanvasColorTables.DelphiRound(
            (srcRect.Right - srcRect.Left) / (double)(max - min) * Math.Min(value, max));
        return srcRect;
    }

    /// <summary>
    /// DxImageProgress.pas 284-291 的值文本 1:1（含原文的分支顺序与遗漏）：
    ///   vtValueAndMax → Prefix + Value + Splite + Max + Suffix
    ///   vtValue       → Prefix + Value + Suffix
    ///   vtPercentage  → Prefix + Format('%d%%', [Round(Value / (Max - Min) * 100)]) + Suffix
    ///   其余（**含 vtNone**）→ ''（空串 → 调用方因此不绘制）
    /// </summary>
    public static string ComputeValueText(TProgressValueType valueType, string prefix, string suffix,
        string splite, uint max, uint min, uint value)
    {
        if (valueType == TProgressValueType.vtValueAndMax)
            return prefix + value + splite + max + suffix;
        if (valueType == TProgressValueType.vtValue)
            return prefix + value + suffix;
        if (valueType == TProgressValueType.vtPercentage)
            return prefix + TDxCanvasColorTables.DelphiRound(value / (double)(max - min) * 100) + "%" + suffix;
        return "";
    }

    /// <summary>
    /// DxImageProgress.pas 350-366 RecallAutoSize 1:1。
    /// AutoSize 为假 → 直接 Exit；仅当 ImageBG &gt;= 0 且取到图且 **D.Width * D.Height &gt; 4** 才定尺。
    /// </summary>
    public void RecallAutoSize()
    {
        if (!AutoSize) return;

        if (ProgressSetting.Image != null)
        {
            if (ProgressSetting.ImageBG >= 0)
            {
                var d = ProgressSetting.Image.GetImage(ProgressSetting.ImageBG);
                if (d != null && d.Width * d.Height > 4)
                {
                    Width = d.Width;
                    Height = d.Height;
                }
            }
        }
    }

    /// <summary>DxImageProgress.pas 368-400 Assign 1:1。</summary>
    public void Assign(TDxControl source)
    {
        if (source is TDxImageProgress src)
        {
            Left = src.Left;
            Top = src.Top;
            Width = src.Width;
            Height = src.Height;
            Enabled = src.Enabled;
            Visible = src.Visible;

            Transparent = src.Transparent;
            EnableFocus = src.EnableFocus;
            Floating = src.Floating;
            OwnerMove = src.OwnerMove;

            MouseEvents = src.MouseEvents;

            ReferenceX = src.ReferenceX;
            AdjustYByHeight = src.AdjustYByHeight;
            TopAlignment = src.TopAlignment;

            Center = src.Center;
            Align = src.Align;
            AutoSize = src.AutoSize;
            BackgroundColor = src.BackgroundColor;

            ImageIndex.Assign(src.ImageIndex);
            BorderColor.Assign(src.BorderColor);

            ProgressSetting.Assign(src.ProgressSetting);
        }
    }
}

/// <summary>TDxImageProgress.PaintTo 的绘制结果（headless 断言用）。</summary>
public sealed class TDxImageProgressPaintResult
{
    public bool DrewBackground;
    public bool DrewProgress;
    public bool DrewValueText;
    public bool DrewDesignFrame;
    public TDxRect BackgroundSrcRect;
    public TDxRect ProgressSrcRect;
    public string ValueText = "";
}
